// @sudoku/stores/domainGame.js
/**
 * Svelte store adapter for the domain Game.
 *
 * Acts as the bridge between the pure-JS domain objects (Sudoku / Game)
 * and the Svelte reactive system.  It exposes:
 *
 *   subscribe   – Svelte-compatible subscriber (current 9×9 grid)
 *   canUndo     – readable store (boolean)
 *   canRedo     – readable store (boolean)
 *   init(grid)  – start a new game with the given puzzle
 *   guess(cursor, value) – forward a user move into the domain Game
 *   undo()      – undo last move
 *   redo()      – redo last undone move
 *   getGame()   – access the underlying Game instance (e.g. for hints)
 */
import { writable } from 'svelte/store';
import { createGame, createSudoku } from '../../../domain/index.js';

function createDomainGameStore() {
	let _game = null;

	// Internal writable stores for reactive state
	const _grid    = writable(null);
	const _canUndo = writable(false);
	const _canRedo = writable(false);

	/** Push the current game state into all reactive stores */
	function _notify() {
		if (!_game) return;
		_grid.set(_game.getSudoku().getGrid());
		_canUndo.set(_game.canUndo());
		_canRedo.set(_game.canRedo());
	}

	return {
		// Expose the grid as the primary subscribable (mirrors userGrid)
		subscribe: _grid.subscribe,

		// Derived boolean stores so components can disable buttons reactively
		canUndo: { subscribe: _canUndo.subscribe },
		canRedo: { subscribe: _canRedo.subscribe },

		/**
		 * Initialise (or re-initialise) the game with a fresh puzzle.
		 * @param {number[][]} puzzle  9×9 grid (0 = empty)
		 */
		init(puzzle) {
			_game = createGame({ sudoku: createSudoku(puzzle) });
			_notify();
		},

		/**
		 * Record a user move in the domain Game and notify Svelte.
		 * @param {{ x: number, y: number }} cursor  Cell position
		 * @param {number} value
		 */
		guess(cursor, value) {
			if (!_game) return;
			_game.guess({ row: cursor.y, col: cursor.x, value });
			_notify();
		},

		/** Undo the last move */
		undo() {
			if (!_game) return;
			_game.undo();
			_notify();
		},

		/** Redo the last undone move */
		redo() {
			if (!_game) return;
			_game.redo();
			_notify();
		},

		/** Direct access to the underlying Game (e.g. for hint computation) */
		getGame() {
			return _game;
		},
	};
}

export const domainGame = createDomainGameStore();
