// @sudoku/stores/domainGame.js
/**
 * Svelte store adapter for the domain Game.
 *
 * This is the bridge between the pure domain objects (Sudoku / Game) and the
 * Svelte reactive system.  The adapter:
 *  - Internally holds a `Game` domain object.
 *  - Exposes Svelte-readable stores for `userGrid`, `canUndo`, `canRedo`.
 *  - Exposes methods (`init`, `guess`, `applyHint`, `undo`, `redo`) that the
 *    UI calls, which update the domain object and then push the new state into
 *    the stores so Svelte re-renders.
 *
 * Why this works:
 *  Svelte tracks subscriptions to stores created with `writable()`.  Every
 *  time we call `store.set(newValue)` the value reference changes, which
 *  triggers re-renders in any component using `$store`.  We intentionally
 *  replace the entire grid array (not mutate in place) so Svelte always sees
 *  a new reference and reliably rerenders.
 */
import { writable } from 'svelte/store';
import { createGame, createSudoku } from '../../../domain/index.js';

function emptyGrid() {
	return Array.from({ length: 9 }, () => Array(9).fill(0));
}

function createDomainGameStore() {
	let _game = null;

	// Reactive state exposed to the UI
	const _userGrid = writable(emptyGrid());
	const _canUndo = writable(false);
	const _canRedo = writable(false);

	/**
	 * Push the current domain state into the reactive stores.
	 * Called after every mutation so Svelte sees a new reference.
	 */
	function _sync() {
		_userGrid.set(_game.getSudoku().getGrid());
		_canUndo.set(_game.canUndo());
		_canRedo.set(_game.canRedo());
	}

	return {
		/** Readable store: the current 9×9 board (with user guesses). */
		userGrid: { subscribe: _userGrid.subscribe },

		/** Readable store: whether undo is available. */
		canUndo: { subscribe: _canUndo.subscribe },

		/** Readable store: whether redo is available. */
		canRedo: { subscribe: _canRedo.subscribe },

		/**
		 * Initialize a new game from a puzzle grid.
		 * @param {number[][]} puzzleGrid
		 */
		init(puzzleGrid) {
			_game = createGame({ sudoku: createSudoku(puzzleGrid) });
			_sync();
		},

		/**
		 * Place a value at a position.
		 * @param {{ x: number, y: number }} pos - cursor position (x=col, y=row)
		 * @param {number} value
		 */
		guess(pos, value) {
			if (_game) {
				_game.guess({ row: pos.y, col: pos.x, value });
				_sync();
			}
		},

		/**
		 * Apply the solved value at a position (hint).
		 * @param {{ x: number, y: number }} pos
		 * @param {number} hintValue - the correct value
		 */
		applyHint(pos, hintValue) {
			if (_game) {
				_game.guess({ row: pos.y, col: pos.x, value: hintValue });
				_sync();
			}
		},

		/** Undo the last move. */
		undo() {
			if (_game) {
				_game.undo();
				_sync();
			}
		},

		/** Redo the last undone move. */
		redo() {
			if (_game) {
				_game.redo();
				_sync();
			}
		},

		/** Return the underlying domain Game (e.g. for serialization). */
		getGame() {
			return _game;
		},
	};
}

export const domainGame = createDomainGameStore();
